### PR TITLE
feat: Add ABR startup quality attributes

### PR DESCRIPTION
**Pull request by jondahl**
## Summary

Adds three new opt-in HTML attributes to improve ABR startup quality across all Mux media elements:

- **`initial-bandwidth-estimate-kbps`** — Seeds the HLS.js EWMA bandwidth estimator instead of relying on TCP slow start, which poisons the first measurement and causes segments 2-5 to drop to low renditions.
- **`initial-estimate-segments`** — Number of segments that use the initial estimate before switching to measured bandwidth. Setting to `3` means the first three segments use the initial estimate, segment four is the first to use real data.
- **`min-preload-segments`** — Buffers N segments before allowing playback to start (autoplay or user-initiated). Gives ABR real bandwidth data and builds buffer runway before playback begins.

No defaults changed. All three attributes are opt-in and must be explicitly set.

### Packages updated

| Package | Changes |
|---------|---------|
| `playback-core` | Core implementation: HLS.js config seeding, estimator reset logic, autoplay gating with play/pause interception |
| `mux-video` | Attribute constants, getter/setter properties |
| `mux-audio` | Attribute constants, getter/setter properties |
| `mux-player` | Getter/setters, `getProps()`, template forwarding to `<mux-video>` |
| `mux-video-react` | PropTypes entries |
| `mux-audio-react` | PropTypes entries |
| `mux-player-react` | TypeScript type definitions in `MuxMediaPropTypes` |
| `mux-player-astro` | TypeScript type definitions in `MuxPlayerProps` |

### Key implementation details

- `initialBandwidthEstimateKbps` sets `abrEwmaDefaultEstimate` in HLS.js config via conditional spread (no `Record<string, any>` type widening)
- `initialEstimateSegments` resets the EWMA estimator after each of the first N-1 segments via `FRAG_BUFFERED` events
- `minPreloadSegments` gates playback in `setupAutoplay()` by intercepting play events, pausing immediately, and resuming once enough segments buffer. Handles both autoplay and user-initiated play, with proper cleanup of event listeners

## Test Coverage

- 6 new tests for `initial-bandwidth-estimate-kbps` (3 tests) and `initial-estimate-segments` (3 tests)
- Tests use `hls.trigger()` with `stats: { aborted: true }, elementaryStreams: {}` stubs to avoid HLS.js internal handler crashes
- All 83 playback-core tests pass

## Reference docs updated

- `packages/mux-video/README.md`
- `packages/mux-player/REFERENCE.md`
- `packages/mux-player-react/REFERENCE.md`
- `packages/mux-player-astro/REFERENCE.md`

## Test plan

- [x] All 83 playback-core tests pass (0 failures)
- [x] TypeScript compiles clean across playback-core, mux-video, mux-audio, mux-player, mux-player-react
- [x] Manual testing with `abr-startup-test.html` harness (included in examples/)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches core playback/ABR behavior (HLS.js config and fragment buffering event handling) and manipulates `playbackRate` during startup, which could affect playback edge-cases, though changes are opt-in and covered by new tests.
> 
> **Overview**
> Adds three new **opt-in** ABR startup controls across Mux media elements: `initial-bandwidth-estimate-kbps`, `initial-estimate-segments`, and `min-preload-segments`.
> 
> In `@mux/playback-core`, HLS.js initialization can now seed `abrEwmaDefaultEstimate` from `initialBandwidthEstimateKbps`, reset the EWMA estimator for the first N buffered segments (`initialEstimateSegments`), and delay visible playback until N main segments buffer by clamping `mediaEl.playbackRate` to `0` (`minPreloadSegments`).
> 
> Plumbs these attributes through `mux-video`, `mux-audio`, and `mux-player` (props extraction + template forwarding), updates React/Astro typings/docs, adds playback-core tests for estimate seeding/reset behavior, and includes a new `abr-startup-test.html` manual test harness.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 960f52f024edd8121038e4c7f569d5b2ba93ad95. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/playback-core/src/autoplay.ts [hunk 1]
```diff
@@ -21,11 +21,65 @@ export const setupAutoplay = (
...
+  }
+
+  const maybeHandleAutoplay = (el: HTMLMediaElement, ap: Autoplay) => {
+    if (preloadReady) {
+      handleAutoplay(el, ap);
+    } else if (ap) {
+      pendingAutoplay = true;
+    }
+  };
+
```
Comment (luwes) [inline packages/playback-core/src/autoplay.ts:82]: Everything looks good but this feature of minPreloadSegments is not what we want I think. This just defers playback and still keeps the segment renditions in place what it chose initially I'm assuming.

If we would delay playback I imagined we or the hls.js engine would more intelligently choose segments and possibly overwrite the buffer with better quality ones.

Unless I'm misunderstanding it I would remove the minPreloadSegments feature for something that actually overwrites the buffer.

**Review by luwes — CHANGES_REQUESTED** [review]
Questioning `minPreloadSegments`

File: packages/playback-core/src/autoplay.ts (hunk 1 above)
Comment (jondahl) [reply packages/playback-core/src/autoplay.ts:82]: the intention behind this is to (optionally) to let a user delay playback until the "forced size" segments are buffered. Example:

* I set initial bandwidth to 10 Mbps, for the first 4 segments
* User has 1 Mbps
* Without minPreloadSegments: buffering between each of those 4 segments?
* With minPreloadSegments: slow loading of the first 4 segments, which will the play back seamlessly, with a downshift to ~1 Mbps for segment 5
* Intention: give customer control over this tradeoff. (Do you prefer a low bandwidth user has slow startup + smooth playback, or fast startup with potential rebuffering?)

If there is a better way to do this, 100% fine with me.

**Review by luwes — APPROVED** [review]
LGTM! I do think the API needs some more rethinking to make it more developer friendly. I'd like to help with that in Video.js.

**Comment by luwes** [comment]
actually, we need to trigger a manual waiting event for the minPreloadSegments. looking into it

File: packages/playback-core/src/min-preload.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  let userPlaybackRate = mediaEl.playbackRate || 1;
+  // Guards `ratechange` against our own internal writes so we only
+  // capture user-driven rate changes.
+  let internalRateUpdate = false;
+
+  const setRateInternal = (rate: number) => {
+    internalRateUpdate = true;
+    mediaEl.playbackRate = rate;
+    internalRateUpdate = false;
+  };
```
[the remark this answers is not in the document]
Comment (luwes) [reply packages/playback-core/src/min-preload.ts]: this ain't true :/
